### PR TITLE
patch before and beforeEach methods

### PR DESCRIPTION
**Pull request by tchambard**
@bjouhier Hi Bruno,
Could you have a look to these changes please ?

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,13 +1,30 @@
...
 
-export function setup() {
-    const glob = global as any;
-    const it = glob.it;
-    if (it.wrapped) return;
-    glob.it = function (name: string, fn: () => void) {
-        return it(name, function (done: MochaDone) {
+
+function wrapWithRun(_fn: Function) {
+    return function (name: string, fn: () => void) {
```
Comment (bjouhier) [inline src/index.ts]: I would rename `fn` -> `body` and `_fn` -> `fn`

File: test/basic-test.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
-import { assert } from 'chai';
-import { wait } from 'f-promise';
-import { setup } from '..';
+import {assert} from 'chai';
+import {wait} from 'f-promise';
+import {setup} from '..';
```
Comment (bjouhier) [inline test/basic-test.ts]: Why the reformatting here?

File: src/index.ts [hunk 3]
```diff
@@ -1,13 +1,30 @@
...
+        }
+        return _fn(name, function (done: MochaDone) {
             run(() => fn()).then(done, done);
         });
     }
-    glob.it.wrapped = true;
+}
+export function setup() {
+    function patchFn(fnName: string, keep?: string[]) {
+        keep = keep || [];
```
Comment (bjouhier) [inline src/index.ts]: I'd rename `keep` -> `subNames`. 
Also you could set the default value in the parameter declaration: `subNames: string[] = []`

File: test/basic-test.ts [hunk 4]
```diff
@@ -13,4 +13,79 @@ describe("basic test", () => {
...
+        });
+
+        it('check expected value set in beforeEach but changed in previous it', () => {
+            assert(expected, 'got expected value set in before');
+        });
+    });
+    describe("async test", () => {
+        let expected: boolean;
+        beforeEach('init expected value', () => {
+            assert(true, 'got false before wait');
```
Comment (bjouhier) [inline test/basic-test.ts]: see previous comment

File: src/index.ts (hunk 1 above)
Comment (tchambard) [inline src/index.ts]: ok

File: src/index.ts (hunk 3 above)
Comment (tchambard) [inline src/index.ts]: ok for renaming, but default value should stay like that to keep compliant with Node 4

File: test/basic-test.ts (hunk 2 above)
Comment (tchambard) [inline test/basic-test.ts]: Webstorm did it automatically, I will revert it

**Comment by tchambard** [comment]
@bjouhier requested changes applied.

**Pull request merged**